browser-extension/mv3/src/page-scripts/ajaxRequestInterceptor/fetch.js

```diff
@@ -0,0 +1,225 @@
+        ruleDetails: requestRule,
+        requestDetails: {
+          url: url,
+          method: request.method,
+          type: "fetch",
+          timeStamp: Date.now(),
+        },
+      });
+    }
+
+    let requestData;
+    if (canRequestBodyBeSent) {
+      requestData = jsonifyValidJSONString(await request.clone().text()); // cloning because the request will be used to make API call
+    }
+
+    const responseRule = getMatchedResponseRule({
+      url,
+      requestData,
+      method,
+    });
+
+    if (responseRule) {
+      console.log("[fetchInterceptor] matchedResponseRule", { responseRule });
+    }
+
+    let responseHeaders;
+    let fetchedResponse;
+
+    if (responseRule && shouldServeResponseWithoutRequest(responseRule)) {
+      const contentType = isJSON(responseRule.pairs[0].response.value) ? "application/json" : "text/plain";
+      responseHeaders = new Headers({ "content-type": contentType });
+    } else {
+      try {
+        const headersObject = {};
+        request?.headers?.forEach((value, key) => {
+          headersObject[key] = value;
+        });
+        await notifyOnBeforeRequest({
+          url,
+          method,
+          type: "xmlhttprequest",
+          initiator: location.origin,
+          requestHeaders: headersObject,
+        });
+
+        if (requestRule) {
+          // use modified request to fetch response
+          fetchedResponse = await _fetch(request);
+        } else {
+          fetchedResponse = await getOriginalResponse();
+        }
+
+        if (!responseRule) {
+          return fetchedResponse;
+        }
+
+        responseHeaders = fetchedResponse?.headers;
+      } catch (error) {
+        if (!responseRule) {
+          return Promise.reject(error);
+        }
+      }
+    }
+
+    debug &&
+      console.log("RQ", "Inside the fetch block for url", {
+        url,
+        resource,
+        initOptions,
+        fetchedResponse,
+      });
+
+    let customResponse;
+    const responseModification = responseRule.pairs[0].response;
+
+    if (responseModification.type === "code") {
+      const requestHeaders =
+        request.headers &&
+        Array.from(request.headers).reduce((obj, [key, val]) => {
+          obj[key] = val;
+          return obj;
+        }, {});
+
+      let evaluatorArgs = {
+        method,
+        url,
+        requestHeaders,
+        requestData,
+      };
+
+      if (fetchedResponse) {
+        const fetchedResponseData = await fetchedResponse.text();
+        const responseType = fetchedResponse.headers.get("content-type");
+        const fetchedResponseDataAsJson = jsonifyValidJSONString(fetchedResponseData, true);
+
+        evaluatorArgs = {
+          ...evaluatorArgs,
+          responseType,
+          response: fetchedResponseData,
+          responseJSON: fetchedResponseDataAsJson,
+        };
+      }
+
+      customResponse = getFunctionFromCode(responseModification.value)(evaluatorArgs);
```

Comment:
If there's error in `getFunctionFromCode` and `undefined` is returned from it then `undefined` shouldn't be returned as API's response instead the original response should be returned